RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorGPU.cc

```diff
@@ -141,243 +109,127 @@ void CAHitQuadrupletGeneratorGPU::fillResults(
 
-  // Riemann-Fit related allocations
-  cudaCheck(cudaMalloc(&hitsGPU_, 48 * maxNumberOfQuadruplets_ * sizeof(Rfit::Matrix3xNd(3, 4))));
-  cudaCheck(cudaMemset(hitsGPU_, 0x00, 48 * maxNumberOfQuadruplets_ * sizeof(Rfit::Matrix3xNd(3, 4))));
+  //product
+  cudaCheck(cudaMalloc(&gpu_.tuples_d, sizeof(TuplesOnGPU::Container)));
+  cudaCheck(cudaMalloc(&gpu_.apc_d, sizeof(AtomicPairCounter)));
+  cudaCheck(cudaMalloc(&gpu_.helix_fit_results_d, sizeof(Rfit::helix_fit)*maxNumberOfQuadruplets_));
+  cudaCheck(cudaMalloc(&gpu_.quality_d, sizeof(Quality)*maxNumberOfQuadruplets_));
 
-  cudaCheck(cudaMalloc(&hits_covGPU_, 48 * maxNumberOfQuadruplets_ * sizeof(Rfit::Matrix3Nd(12, 12))));
-  cudaCheck(cudaMemset(hits_covGPU_, 0x00, 48 * maxNumberOfQuadruplets_ * sizeof(Rfit::Matrix3Nd(12, 12))));
+  cudaCheck(cudaMalloc(&gpu_d, sizeof(TuplesOnGPU)));
+  gpu_.me_d = gpu_d;
+  cudaCheck(cudaMemcpy(gpu_d, &gpu_, sizeof(TuplesOnGPU), cudaMemcpyDefault));
 
-  cudaCheck(cudaMalloc(&fast_fit_resultsGPU_, 48 * maxNumberOfQuadruplets_ * sizeof(Eigen::Vector4d)));
-  cudaCheck(cudaMemset(fast_fit_resultsGPU_, 0x00, 48 * maxNumberOfQuadruplets_ * sizeof(Eigen::Vector4d)));
+  cudaCheck(cudaMallocHost(&tuples_, sizeof(TuplesOnGPU::Container)));
+  cudaCheck(cudaMallocHost(&helix_fit_results_, sizeof(Rfit::helix_fit)*maxNumberOfQuadruplets_));
+  cudaCheck(cudaMallocHost(&quality_, sizeof(Quality)*maxNumberOfQuadruplets_));
 
-  cudaCheck(cudaMalloc(&circle_fit_resultsGPU_, 48 * maxNumberOfQuadruplets_ * sizeof(Rfit::circle_fit)));
-  cudaCheck(cudaMemset(circle_fit_resultsGPU_, 0x00, 48 * maxNumberOfQuadruplets_ * sizeof(Rfit::circle_fit)));
+  kernels.allocateOnGPU();
+  fitter.allocateOnGPU(gpu_.tuples_d, gpu_.helix_fit_results_d);
 
-  cudaCheck(cudaMalloc(&line_fit_resultsGPU_, maxNumberOfQuadruplets_ * sizeof(Rfit::line_fit)));
-  cudaCheck(cudaMemset(line_fit_resultsGPU_, 0x00, maxNumberOfQuadruplets_ * sizeof(Rfit::line_fit)));
 
-  cudaCheck(cudaMalloc(&helix_fit_resultsGPU_, sizeof(Rfit::helix_fit)*maxNumberOfQuadruplets_));
-  cudaCheck(cudaMemset(helix_fit_resultsGPU_, 0x00, sizeof(Rfit::helix_fit)*maxNumberOfQuadruplets_));
 }
 
-void CAHitQuadrupletGeneratorGPU::deallocateOnGPU()
+void CAHitQuadrupletGeneratorGPU::launchKernels(HitsOnCPU const & hh,
+                                                bool doRiemannFit,
+                                                bool transferToCPU,
+                                                cudaStream_t cudaStream)
 {
-  for (size_t i = 0; i < h_foundNtupletsVec_.size(); ++i)
-  {
-    cudaFreeHost(h_foundNtupletsVec_[i]);
-    cudaFreeHost(h_foundNtupletsData_[i]);
-    cudaFree(d_foundNtupletsVec_[i]);
-    cudaFree(d_foundNtupletsData_[i]);
+
+  kernels.launchKernels(hh, gpu_, cudaStream); 
+  if (doRiemannFit) {
+    fitter.launchKernels(hh, hh.nHits, CAConstants::maxNumberOfQuadruplets(), cudaStream);
+    kernels.classifyTuples(hh, gpu_, cudaStream);
+  }
+  if (transferToCPU) {
```

Comment:
one of this three (or all of them) seems to produce a
```
Uninitialized access at 0x2aab2b009dc0 on access by cudaMemcopy source
```
what does it mean?
that we are transferring uninitialized memory?
yes, the memory is not zeroed (why should be?) and we fill much less tracks then ```CAConstants::maxNumberOfQuadruplets()```
initializing memory for each event is out of discussion.
If silencing a false positive is a requirement we can memset these storage area after allocation

Reply:
>  that we are transferring uninitialized memory?

I think so, yes.

>  initializing memory for each event is out of discussion.

Agreed.

> If silencing a false positive is a requirement ...

Making the tool happy is not a requirement, but I would say that silencing the error report is one - if only to be able to find the eventual real problems in the future.

> ... we can memset these storage area after allocation.

Sounds good.

Reply:
initchk silenced here https://github.com/cms-patatrack/cmssw/commit/c46e71620eecf7c25387c09e1b037c35da1396e3
(on top of #216 and #236)